File: Soham_Mukherjee/Task-01/duckduckgo.go
//  running instructions: go run duckduckgo.go Narendra Modi := search_Results for Narendra Modi

package main

import (
	"encoding/json"
	"fmt"
	"log"
	"net/http"
	"net/url"
	"os"
	"strings"
	"time"
)

var ddgURL = "https://api.duckduckgo.com/"

type NetFail struct {
	Err error
}

func (e NetFail) Error() string {
	return fmt.Sprintf("netfail: %v", e.Err)
}

func httpGetJSON(endpoint string, params map[string]string, retries int, timeout time.Duration) (map[string]interface{}, error) {
	var last error
	client := &http.Client{Timeout: timeout}
	values := url.Values{}
	for k, v := range params {
		values.Set(k, v)
	}
	fullURL := endpoint + "?" + values.Encode()
	for i := 0; i < retries; i++ {
		resp, err := client.Get(fullURL)
		if err != nil {
			last = err
			log.Printf("retry %d/%d: %v", i+1, retries, err)
			time.Sleep(time.Duration(1<<i)*time.Second + time.Millisecond*500)
			continue
		}
		defer resp.Body.Close()
		if resp.StatusCode >= 500 {
			last = fmt.Errorf("bad upstream %d", resp.StatusCode)
			log.Printf("retry %d/%d: %v", i+1, retries, last)
			time.Sleep(time.Duration(1<<i)*time.Second + time.Millisecond*500)
			continue
		}
		var result map[string]interface{}
		if err := json.NewDecoder(resp.Body).Decode(&result); err != nil {
			last = err
			log.Printf("retry %d/%d: %v", i+1, retries, err)
			time.Sleep(time.Duration(1<<i)*time.Second + time.Millisecond*500)
			continue
		}
		return result, nil
	}
	return nil, NetFail{last}
}

func ddgParams(q string) map[string]string {
	return map[string]string{
		"q":             q,
		"format":        "json",
		"no_html":       "1",
		"skip_disambig": "1",
		"t":             "agent",
	}
}

func ddg(q string) (map[string]interface{}, error) {
	return httpGetJSON(ddgURL, ddgParams(q), 3, 8*time.Second)
}

func flatten(rt []interface{}) []map[string]string {
	out := []map[string]string{}
	for _, item := range rt {
		m, ok := item.(map[string]interface{})
		if !ok {
			continue
		}
		if topics, ok := m["Topics"].([]interface{}); ok {
			for _, t := range topics {
				topic, ok := t.(map[string]interface{})
				if ok && topic["FirstURL"] != nil && topic["Text"] != nil {
					out = append(out, map[string]string{
						"title": topic["Text"].(string),
						"url":   topic["FirstURL"].(string),
					})
				}
			}
		} else {
			if m["FirstURL"] != nil && m["Text"] != nil {
				out = append(out, map[string]string{
					"title": m["Text"].(string),
					"url":   m["FirstURL"].(string),
				})
			}
		}
	}
	seen := map[string]bool{}
	dedup := []map[string]string{}
	for _, x := range out {
		if !seen[x["url"]] {
			seen[x["url"]] = true
			dedup = append(dedup, x)
		}
	}
	return dedup
}

func pick(d map[string]interface{}) map[string]string {
	if v, ok := d["Answer"].(string); ok && v != "" {
		return map[string]string{"kind": "answer", "value": v}
	}
	if v, ok := d["Definition"].(string); ok && v != "" {
		return map[string]string{"kind": "definition", "value": v}
	}
	if v, ok := d["AbstractText"].(string); ok && v != "" {
		return map[string]string{"kind": "abstract", "value": v}
	}
	if h, ok := d["Heading"].(string); ok && h != "" {
		if a, ok := d["Abstract"].(string); ok && a != "" {
			return map[string]string{"kind": "abstract", "value": a}
		}
	}
	return nil
}

func search(q string, want int) map[string]interface{} {
	log.Printf("searching: %s", q)
	raw, err := ddg(q)
	if err != nil {
		log.Printf("ddg err: %v", err)
		raw = map[string]interface{}{}
	}
	var related []map[string]string
	if rt, ok := raw["RelatedTopics"].([]interface{}); ok {
		related = flatten(rt)
	}
	ans := pick(raw)
	if ans != nil {
		return map[string]interface{}{
			"query":   q,
			"mode":    "instant",
			"answer":  ans,
			"related": related[:min(want, len(related))],
		}
	}
	if len(related) > 0 {
		return map[string]interface{}{
			"query":   q,
			"mode":    "related_only",
			"related": related[:min(want, len(related))],
			"hint":    "try broader search",
		}
	}
	return map[string]interface{}{
		"query":   q,
		"mode":    "dry",
		"related": related[:min(want, len(related))],
		"hint":    "nothing solid",
	}
}

func asText(r map[string]interface{}) string {
	lines := []string{}
	if ans, ok := r["answer"].(map[string]string); ok {
		lines = append(lines, fmt.Sprintf("%s: %s", ans["kind"], ans["value"]))
	}
	if related, ok := r["related"].([]map[string]string); ok && len(related) > 0 {
		lines = append(lines, "related:")
		for _, x := range related {
			lines = append(lines, fmt.Sprintf("- %s -> %s", x["title"], x["url"]))
		}
	}
	if hint, ok := r["hint"].(string); ok && hint != "" {
		lines = append(lines, hint)
	}
	return strings.Join(lines, "\n")
}

func min(a, b int) int {
	if a < b {
		return a
	}
	return b
}

func main() {
	if len(os.Args) < 2 {
		fmt.Println("usage: duckduckgo <query> [--json] [--n N]")
		os.Exit(1)
	}
	args := os.Args[1:]
	asJSON := false
	want := 6
	queryParts := []string{}
	for i := 0; i < len(args); i++ {
		if args[i] == "--json" {
			asJSON = true
		} else if args[i] == "--n" && i+1 < len(args) {
			fmt.Sscanf(args[i+1], "%d", &want)
			i++
		} else {
			queryParts = append(queryParts, args[i])
		}
	}
	q := strings.Join(queryParts, " ")
	r := search(q, want)
	if asJSON {
		b, _ := json.MarshalIndent(r, "", "  ")
		fmt.Println(string(b))
	} else {
		fmt.Println(asText(r))
	}
}

File: Soham_Mukherjee/Task-01/duckduckgo.py
### requirements: requests
### running instructions: python duckduckgo.py


import os, sys, time, json, random, logging, requests, argparse
from dotenv import load_dotenv
from typing import Dict, Any, List

load_dotenv()

logging.basicConfig(level=logging.INFO, format="%(asctime)s - %(levelname)s - %(message)s")
logger = logging.getLogger("ddg")

DDG_URL = os.getenv("DDG_URL")
GCS_URL = os.getenv("GCS_URL")

class netfail(Exception):
    pass

# from langchain_community.utilities import DuckDuckGoSearchAPIWrapper
# wrapper = DuckDuckGoSearchAPIWrapper(region="de-de", time="d", max_results=2)
# search = DuckDuckGoSearchResults(api_wrapper=wrapper, source="news")
# search.invoke("Obama")

def _http(url: str, params: Dict[str, Any], retries: int = 3, timeout: int = 8):
    """make an http get request with retries and timeout"""
    
    last = None
    for i in range(retries):
        try:
            r = requests.get(url, params=params, timeout=timeout)
            if r.status_code >= 500:
                raise netfail(f"bad upstream {r.status_code}")
            r.raise_for_status()
            return r.json()
        
        except Exception as e:
            last = e
            wait = min(2 ** i, 8) + random.random()
            logger.warning(f"retry {i+1}/{retries} in {wait:.1f}s: {e}")
            time.sleep(wait)
    
    raise netfail(f"failed after retries: {last}")

def _ddg_params(q: str):
    """build duckduckgo query parameters"""
    
    return {"q": q, "format": "json", "no_html": 1, "skip_disambig": 1, "t": "agent"}

def ddg(q: str):
    """perform duckduckgo search for the given query"""
    
    return _http(DDG_URL, _ddg_params(q))

def _flatten(rt: List[Dict[str, Any]]):
    """flatten related topics from duckduckgo results"""
    
    out = []
    
    for item in rt or []:
    
        if "Topics" in item:
    
            for t in item.get("Topics", []):
    
                if t.get("FirstURL") and t.get("Text"):
                    out.append({"title": t["Text"], "url": t["FirstURL"]})
    
        else:
    
            if item.get("FirstURL") and item.get("Text"):
                out.append({"title": item["Text"], "url": item["FirstURL"]})
    
    seen, dedup = set(), []
    
    for x in out:
    
        if x["url"] not in seen:
            seen.add(x["url"])
            dedup.append(x)
    
    return dedup

def gcs(q: str, num: int = 5):
    
    """perform google custom search for the given query"""
    
    key, cx = os.getenv("GOOGLE_CSE_KEY"), os.getenv("GOOGLE_CSE_CX")
    
    if not key or not cx:
        logger.info("gcs not configured")
        return []
    
    data = _http(GCS_URL, {"key": key, "cx": cx, "q": q, "num": num})
    
    items = data.get("items", [])
    
    return [{"title": it.get("title"), "url": it.get("link"), "snippet": it.get("snippet")} for it in items]

def _pick(d: Dict[str, Any]):
    """pick the best answer from duckduckgo result dict"""
    if d.get("Answer"): 
        return {"kind": "answer", "value": d["Answer"]}
    
    if d.get("Definition"): 
        return {"kind": "definition", "value": d["Definition"]}
    
    if d.get("AbstractText"): 
        return {"kind": "abstract", "value": d["AbstractText"]}
    
    if d.get("Heading") and d.get("Abstract"): 
        return {"kind": "abstract", "value": d["Abstract"]}
    
    else: 
        return None

def search(q: str, want: int = 6):
    """search using duckduckgo and google custom search, return structured result"""
    
    logger.info(f"searching: {q}")
    
    try:
        raw = ddg(q)
    
    except Exception as e:
        logger.error(f"ddg err: {str(e)}")
        raw = {}
    ans = _pick(raw) if raw else None
    related = _flatten(raw.get("RelatedTopics", [])) if raw else []
    
    if ans:
        return {"query": q, "mode": "instant", "answer": ans, "related": related[:want]}
    
    cse = gcs(q, num=want)
    
    if related and not cse:
        return {"query": q, "mode": "related_only", "related": related[:want], "hint": "try broader search"}
    
    if cse:
        return {"query": q, "mode": "fallback_cse", "related": related[:want], "cse_results": cse}
    
    return {"query": q, "mode": "dry", "related": related[:want], "hint": "nothing solid"}

def as_text(r: Dict[str, Any]):
    """convert search result dict to readable text"""
    
    lines = []
    
    if r.get("answer"): lines.append(f"{r['answer']['kind']}: {r['answer']['value']}")
    
    if r.get("related"):
        lines.append("related:")
        for x in r["related"]:
            lines.append(f"- {x['title']} -> {x['url']}")
    
    if r.get("cse_results"):
        lines.append("web:")
    
        for x in r["cse_results"]:
            lines.append(f"- {x['title']} -> {x['url']}")
    
    if r.get("hint"): lines.append(r["hint"])
    
    return "\n".join(lines)

if __name__ == "__main__":
    p = argparse.ArgumentParser()
    
    p.add_argument("q", nargs="+")
    p.add_argument("--json", dest="as_json", action="store_true")
    p.add_argument("--n", type=int, default=6)
    a = p.parse_args()
    q = " ".join(a.q)
    
    try:
        r = search(q, want=a.n)
    
    except Exception as e:
        logger.error(str(e))
        sys.exit(2)
    
    if a.as_json:
        print(json.dumps(r, ensure_ascii=False))
    
    else:
        print(as_text(r))
